explorer/chartscache.go

```diff
@@ -0,0 +1,403 @@
+package explorer
+
+import (
+	"context"
+	"encoding/gob"
+	"os"
+	"sync"
+
+	"github.com/decred/dcrdata/db/dbtypes"
+	"github.com/decred/dcrdata/txhelpers"
+)
+
+// Cache data for charts that appear on /charts page is managed here.
+
+// chartDataCache is a data cache for the historical charts that appear on
+// /charts page on the UI.
+type chartDataCache struct {
+	sync.RWMutex
+	updateHeight int64
+	Data         map[string]*dbtypes.ChartsData
+}
+
+// cacheChartsData holds the prepopulated data that is used to draw the charts.
+var cacheChartsData chartDataCache
+
+// get provides a thread-safe way to access the all cache charts data.
+func (c *chartDataCache) get() map[string]*dbtypes.ChartsData {
+	c.RLock()
+	defer c.RUnlock()
+	return c.Data
+}
+
+// Height returns the last update height of the charts data cache.
+func (c *chartDataCache) Height() int64 {
+	c.RLock()
+	defer c.RUnlock()
+	return c.height()
+}
+
+// height returns the last update height of the charts data cache. Use Height
+// instead for thread-safe access.
+func (c *chartDataCache) height() int64 {
+	if c.updateHeight == 0 {
+		return -1
+	}
+	return c.updateHeight
+}
+
+// Update sets new data for the given height in the the charts data cache.
+func (c *chartDataCache) Update(height int64, newData map[string]*dbtypes.ChartsData) {
+	c.Lock()
+	defer c.Unlock()
+	c.update(height, newData)
+}
+
+// update sets new data for the given height in the the charts data cache. Use
+// Update instead for thread-safe access.
+func (c *chartDataCache) update(height int64, newData map[string]*dbtypes.ChartsData) {
+	c.updateHeight = height
+	c.Data = newData
+}
+
+// ChartTypeData is a thread-safe way to access chart data of the given type.
+func ChartTypeData(chartType string) (data *dbtypes.ChartsData, ok bool) {
+	cacheChartsData.RLock()
+	defer cacheChartsData.RUnlock()
+
+	// Data updates replace the entire map rather than modifying the data to
+	// which the pointers refer, so the pointer can safely be returned here.
+	data, ok = cacheChartsData.Data[chartType]
```

Comment:
Omitting the bool is better in this case IMO.  The zero value is a `nil` `*dbtypes.ChartsData`.